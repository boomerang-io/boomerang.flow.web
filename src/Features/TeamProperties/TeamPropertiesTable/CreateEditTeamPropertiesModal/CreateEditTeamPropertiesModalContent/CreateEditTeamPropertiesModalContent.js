import React from "react";
import PropTypes from "prop-types";
import { useMutation, queryCache } from "react-query";
import { Formik } from "formik";
import * as Yup from "yup";
import {
  Button,
  InlineNotification,
  ModalBody,
  ModalFooter,
  TextInput,
  Toggle,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { ModalFlowForm, notify, ToastNotification, Loading } from "@boomerang-io/carbon-addons-boomerang-react";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { InputType } from "Constants";
import styles from "./createEditTeamPropertiesModalContent.module.scss";

function CreateEditTeamPropertiesModalContent({ closeModal, isEdit, property, propertyKeys, team, cancelRequestRef }) {
  const teamPropertiesUrl = serviceUrl.getTeamProperties({ id: team.id });

  /** Add Team Property */
  const [addTeamPropertyMutation, { isLoading: addIsLoading, error: addError }] = useMutation(
    (args) => {
      const { promise, cancel } = resolver.postTeamPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(teamPropertiesUrl),
    }
  );

  /** Update Team Property */
  const [updateTeamPropertyMutation, { isLoading: updateIsLoading, error: updateError }] = useMutation(
    (args) => {
      const { promise, cancel } = resolver.patchTeamPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(teamPropertiesUrl),
    }
  );

  const loading = addIsLoading || updateIsLoading;

  const handleSubmit = async (values) => {
    const type = values.secured ? InputType.Password : InputType.Text;
    const newTeamProperty = isEdit ? { ...values, type, id: property.id } : { ...values, type };
    delete newTeamProperty.secured;

    if (isEdit) {
      try {
        const response = await updateTeamPropertyMutation({
          teamId: team.id,
          configurationId: newTeamProperty.id,
          body: newTeamProperty,
        });
        notify(
          <ToastNotification
            kind="success"
            title={"Property Updated"}
            subtitle={`Request to update ${response.data.label} succeeded`}
            data-testid="create-update-team-prop-notification"
          />
        );
        closeModal();
      } catch (err) {}
    } else {
      try {
        const response = await addTeamPropertyMutation({ id: team.id, body: newTeamProperty });
        notify(
          <ToastNotification
            kind="success"
            title={"Property Created"}
            subtitle={`Request to create ${response.data.label} succeeded`}
            data-testid="create-update-team-prop-notification"
          />
        );
        closeModal();
      } catch (err) {
        //no-op
      }
    }
  };

  // Check if key contains alpahanumeric, underscore, dash, and period chars
  const validateKey = (key) => {
    const regexp = /^[a-zA-Z0-9-._]+$/g;
    return regexp.test(key);
  };

  return (
    <Formik
      initialValues={{
        value: property && property.value ? property.value : "",
        key: property && property.key ? property.key : "",
        label: property && property.label ? property.label : "",
        description: property && property.description ? property.description : "",
        secured: property ? property.type === InputType.Password : false,
      }}
      onSubmit={handleSubmit}
      validationSchema={Yup.object().shape({
        label: Yup.string().required("Enter a label"),
        key: Yup.string()
          .required("Enter a key")
          .max(128, "Key must not be greater than 128 characters")
          .notOneOf(propertyKeys || [], "Enter a unique key value for this workflow")
          .test("is-valid-key", "Only alphanumeric, underscore, dash, and period characters allowed", validateKey),
        value: Yup.string().required("Enter a value"),
        description: Yup.string(),
        secured: Yup.boolean(),
      })}
    >
      {(props) => {
        const { values, touched, errors, isValid, handleChange, handleBlur, handleSubmit } = props;

        return (
          <ModalFlowForm onSubmit={handleSubmit}>
            <ModalBody className={styles.formBody}>
              {loading && <Loading />}
              <TextInput
                id="key"
                labelText="Key"
                placeholder="Key"
                name="key"
                value={values.key}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.key && touched.key}
                invalidText={errors.key}
              />
              <TextInput
                id="label"
                labelText="Label"
                placeholder="Label"
                name="label"
                value={values.label}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.label && touched.label}
                invalidText={errors.label}
              />
              <TextInput
                id="description"
                labelText="Description"
                placeholder="Description"
                name="description"
                value={values.description}
                onBlur={handleBlur}
                onChange={handleChange}
              />
              <TextInput
                id="value"
                labelText="Value"
                placeholder="Value"
                name="value"
                value={values.value}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.value && touched.value}
                invalidText={errors.value}
                type={values.secured ? "password" : "text"}
              />
              <Toggle
                id="secured-team-properties-toggle"
                data-testid="secured-team-properties-toggle"
                labelText="Secured"
                name="secured"
                onChange={handleChange}
                toggled={values.secured}
              />
              {addError && (
                <InlineNotification
                  kind="error"
                  subtitle={"Request to create property failed"}
                  title={"Something's Wrong"}
                  data-testid="create-update-team-prop-notification"
                />
              )}
              {updateError && (
                <InlineNotification
                  kind="error"
                  lowContrast
                  subtitle={"Request to update property failed"}
                  title={"Something's Wrong"}
                  data-testid="create-update-team-prop-notification"
                />
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" type="button" onClick={closeModal}>
                Cancel
              </Button>
              <Button
                data-testid="team-property-create-edit-submission-button"
                type="submit"
                disabled={!isValid || loading}
              >
                {isEdit ? (loading ? "Saving..." : "Save") : loading ? "Creating..." : "Create"}
              </Button>
            </ModalFooter>
          </ModalFlowForm>
        );
      }}
    </Formik>
  );
}

CreateEditTeamPropertiesModalContent.propTypes = {
  isEdit: PropTypes.bool,
  property: PropTypes.object,
  propertyKeys: PropTypes.array.isRequired,
  team: PropTypes.object.isRequired,
  cancelRequestRef: PropTypes.object.isRequired,
};

export default CreateEditTeamPropertiesModalContent;
