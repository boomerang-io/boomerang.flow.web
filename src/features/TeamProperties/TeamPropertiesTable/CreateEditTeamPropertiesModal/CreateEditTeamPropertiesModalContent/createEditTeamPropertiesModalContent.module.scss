@import "~Styles/variables";

.input {
  position: relative;
  height: 6rem;
}

.toggleContainer {
  margin-bottom: 1rem;
}
