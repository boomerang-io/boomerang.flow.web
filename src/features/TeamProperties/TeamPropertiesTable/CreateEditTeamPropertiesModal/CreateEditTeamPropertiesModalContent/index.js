import React, { Component } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Formik } from "formik";
import * as Yup from "yup";
import { Button, ModalBody, ModalFooter, TextInput, Toggle } from "carbon-components-react";
import { ModalFlowForm, notify, ToastNotification } from "@boomerang/carbon-addons-boomerang-react";
import Loading from "Components/Loading";
import { TEAM_PROPERTIES_ID_URL, TEAM_PROPERTIES_ID_PROPERTY_ID_URL } from "Config/servicesConfig";
import INPUT_TYPES from "Constants/inputTypes";
import styles from "./createEditTeamPropertiesModalContent.module.scss";

class CreateEditTeamPropertiesModalContent extends Component {
  static propTypes = {
    addTeamPropertyInStore: PropTypes.func,
    closeModal: PropTypes.func,
    handleEditClose: PropTypes.func,
    isEdit: PropTypes.bool,
    property: PropTypes.object,
    propertyKeys: PropTypes.array.isRequired,
    team: PropTypes.string.isRequired,
    updateTeamProperty: PropTypes.func,
  };

  handleSubmit = (values, options) => {
    const { isEdit, property, addTeamPropertyInStore, updateTeamProperty, team } = this.props;
    const type = values.secured ? INPUT_TYPES.PASSWORD : INPUT_TYPES.TEXT;
    const newTeamProperty = isEdit ? { ...values, type, id: property.id } : { ...values, type };
    const storeUpdate = isEdit ? updateTeamProperty : addTeamPropertyInStore;
    delete newTeamProperty.secured;

    axios({
      method: isEdit ? "patch" : "post",
      url: isEdit ? TEAM_PROPERTIES_ID_PROPERTY_ID_URL(team, newTeamProperty.id) : TEAM_PROPERTIES_ID_URL(team),
      data: newTeamProperty,
    })
      .then((response) => {
        storeUpdate(isEdit ? newTeamProperty : response.data);
        notify(
          <ToastNotification
            kind="success"
            title={isEdit ? "Property Updated" : "Property Created"}
            subtitle={
              isEdit
                ? `Request to update ${response.data.label} succeeded`
                : `Request to create ${response.data.label} succeeded`
            }
            data-testid="create-update-team-prop-notification"
          />
        );
        options.setSubmitting(false);
      })
      .then(() => {
        this.props.closeModal();
      })
      .catch((error) => {
        notify(
          <ToastNotification
            kind="error"
            title={isEdit ? "Update Property Failed" : "Create Property Failed"}
            subtitle={"Something went wrong"}
            data-testid="create-update-team-prop-notification"
          />
        );
        options.setSubmitting(false);
      });
  };

  render() {
    const { isEdit, property, propertyKeys } = this.props;

    return (
      <Formik
        initialValues={{
          value: property && property.value ? property.value : "",
          key: property && property.key ? property.key : "",
          label: property && property.label ? property.label : "",
          description: property && property.description ? property.description : "",
          secured: property ? property.type === INPUT_TYPES.PASSWORD : false,
        }}
        onSubmit={this.handleSubmit}
        validationSchema={Yup.object().shape({
          label: Yup.string().required("Enter a label"),
          key: Yup.string()
            .required("Enter a key")
            .notOneOf(propertyKeys, "Key must be unique"),
          value: Yup.string().required("Enter a value"),
          description: Yup.string(),
          secured: Yup.boolean(),
        })}
        validateOnMount
      >
        {(props) => {
          const { values, touched, errors, isSubmitting, isValid, handleChange, handleBlur, handleSubmit } = props;

          return (
            <ModalFlowForm onSubmit={handleSubmit}>
              <ModalBody className={styles.formBody}>
                {isSubmitting && <Loading />}
                <div className={styles.input}>
                  <TextInput
                    id="label"
                    labelText="Label"
                    name="label"
                    value={values.label}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.label && touched.label}
                    invalidText={errors.label}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="key"
                    labelText="Key"
                    placeholder="Key"
                    name="key"
                    value={values.key}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.key && touched.key}
                    invalidText={errors.key}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="description"
                    labelText="Description"
                    name="description"
                    value={values.description}
                    onBlur={handleBlur}
                    onChange={handleChange}
                  />
                </div>
                <div className={styles.input}>
                  <TextInput
                    id="value"
                    labelText="Value"
                    name="value"
                    value={values.value}
                    onBlur={handleBlur}
                    onChange={handleChange}
                    invalid={errors.value && touched.value}
                    invalidText={errors.value}
                    type={values.secured ? "password" : "text"}
                  />
                </div>
                <div className={styles.toggleContainer}>
                  <Toggle
                    id="secured-team-properties-toggle"
                    data-testid="secured-team-properties-toggle"
                    labelText="Secured"
                    name="secured"
                    onChange={handleChange}
                    toggled={values.secured}
                  />
                </div>
              </ModalBody>
              <ModalFooter>
                <Button kind="secondary" type="button" onClick={this.props.closeModal}>
                  Cancel
                </Button>
                <Button type="submit" disabled={!isValid || isSubmitting}>
                  {isEdit ? (isSubmitting ? "Saving..." : "Save") : isSubmitting ? "Creating..." : "Create"}
                </Button>
              </ModalFooter>
            </ModalFlowForm>
          );
        }}
      </Formik>
    );
  }
}

export default CreateEditTeamPropertiesModalContent;
