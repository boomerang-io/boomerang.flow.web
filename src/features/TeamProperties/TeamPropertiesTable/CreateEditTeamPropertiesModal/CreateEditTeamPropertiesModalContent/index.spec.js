import React from "react";
import CreateEditTeamPropertiesModalContent from "../CreateEditTeamPropertiesModalContent";
import { fireEvent, waitForElement } from "@testing-library/react";

const mockfn = jest.fn();
const props = {
  isEdit: true,
  addTeamPropertyInStore: mockfn,
  updateTeamProperty: mockfn,
  team: "Test-Id",
  propertyKeys: ["key"],
  property: {
    id: 1
  }
};

describe("CreateEditTeamPropertiesModalContent --- Snapshot Test", () => {
  test("Capturing Snapshot of CreateEditTeamPropertiesModalContent", () => {
    const { baseElement } = rtlReduxRender(<CreateEditTeamPropertiesModalContent {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("CreateEditTeamPropertiesModalContent --- RTL Tests", () => {
  test("CreateEditTeamPropertiesModalContent - test if the isActive Toggle appears", () => {
    const newProps = { ...props, isEdit: false };
    const { queryByText } = rtlReduxRender(<CreateEditTeamPropertiesModalContent {...newProps} />);

    expect(queryByText(/active/i)).not.toBeInTheDocument();
  });

  test("CreateEditTeamPropertiesModalContent - test the Submit Button state", () => {
    const newProps = { ...props, isEdit: false };

    const { getByLabelText, findByText } = rtlReduxRender(<CreateEditTeamPropertiesModalContent {...newProps} />);
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);

    waitForElement(() => expect(findByText(/create/i)).toBeDisabled());
    fireEvent.change(valueInputText, { target: { value: "Value Test" } });
    fireEvent.change(labelInputText, { target: { value: "Label Test" } });
    fireEvent.change(keyInputText, { target: { value: "Key Test" } });
    waitForElement(() => expect(findByText(/create/i)).toBeEnabled());
  });

  test("CreateEditTeamPropertiesModalContent - test if the form submits", () => {
    const { getByLabelText, getByText } = rtlReduxRender(<CreateEditTeamPropertiesModalContent {...props} />);
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);
    const saveButton = getByText(/save/i);

    expect(valueInputText).toBeInTheDocument();
    expect(labelInputText).toBeInTheDocument();
    expect(keyInputText).toBeInTheDocument();

    fireEvent.change(valueInputText, { target: { value: "Value Test" } });
    fireEvent.change(labelInputText, { target: { value: "Label Test" } });
    fireEvent.change(keyInputText, { target: { value: "Key Test" } });
    fireEvent.click(saveButton);

    expect(valueInputText).not.toBeInTheDocument();
    expect(labelInputText).not.toBeInTheDocument();
    expect(keyInputText).not.toBeInTheDocument();
  });

  test("CreateEditTeamPropertiesModalContent - test form reqired validations", async () => {
    const { getByLabelText, findByText, queryByText } = rtlReduxRender(
      <CreateEditTeamPropertiesModalContent {...props} />
    );
    const valueInputText = getByLabelText(/value/i);
    const labelInputText = getByLabelText(/label/i);
    const keyInputText = getByLabelText(/key/i);

    expect(queryByText(/Enter a value/i)).toBeNull();
    fireEvent.change(valueInputText, { target: { value: "" } });
    fireEvent.blur(valueInputText);
    const mandatoryValueErr = await findByText(/Enter a value/i);
    expect(mandatoryValueErr).toBeInTheDocument();

    expect(queryByText(/Enter a label/i)).toBeNull();
    fireEvent.change(labelInputText, { target: { value: "" } });
    fireEvent.blur(labelInputText);
    const mandatoryLabelErr = await findByText(/Enter a label/i);
    expect(mandatoryLabelErr).toBeInTheDocument();

    expect(queryByText(/Enter a key/i)).toBeNull();
    fireEvent.change(keyInputText, { target: { value: "" } });
    fireEvent.blur(keyInputText);
    const mandatoryKeyErr = await findByText(/Enter a key/i);
    expect(mandatoryKeyErr).toBeInTheDocument();
  });
});
