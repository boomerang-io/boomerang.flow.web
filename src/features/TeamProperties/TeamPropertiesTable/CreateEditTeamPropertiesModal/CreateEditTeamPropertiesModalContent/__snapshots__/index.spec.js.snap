// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditTeamPropertiesModalContent --- Snapshot Test Capturing Snapshot of CreateEditTeamPropertiesModalContent 1`] = `
<body
  id="app"
>
  <div>
    <form
      class="bx--bmrg-modal-flow-form"
    >
      <div
        class="bx--modal-content formBody"
      >
        <div
          class="input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="label"
            >
              Label
            </label>
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="label"
                name="label"
                placeholder="Label"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="key"
            >
              Key
            </label>
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="key"
                name="key"
                placeholder="Key"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="description"
            >
              Description
            </label>
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="description"
                name="description"
                placeholder="Description"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="value"
            >
              Value
            </label>
            <div
              class="bx--text-input__field-wrapper"
            >
              <input
                class="bx--text-input bx--text__input"
                id="value"
                name="value"
                placeholder="Value"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="toggleContainer"
        >
          <div
            class="bx--form-item"
          >
            <input
              class="bx--toggle-input"
              data-testid="secured-team-properties-toggle"
              id="secured-team-properties-toggle"
              name="secured"
              type="checkbox"
            />
            <label
              class="bx--toggle-input__label"
              for="secured-team-properties-toggle"
            >
              Secured
              <span
                class="bx--toggle__switch"
              >
                <span
                  aria-hidden="true"
                  class="bx--toggle__text--off"
                >
                  Off
                </span>
                <span
                  aria-hidden="true"
                  class="bx--toggle__text--on"
                >
                  On
                </span>
              </span>
            </label>
          </div>
        </div>
      </div>
      <div
        class="bx--modal-footer"
      >
        <button
          class="bx--btn bx--btn--secondary"
          tabindex="0"
          type="button"
        >
          Cancel
        </button>
        <button
          class="bx--btn bx--btn--primary"
          tabindex="0"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</body>
`;
